### PR TITLE
bin cleanup, testing migration script

### PR DESCRIPTION
**Pull request by elrido**
This PR concludes #1007

## Changes
* moves the other two scripts to the new `bin` folder
* renames them all for consistency
* fixes the configuration-test-generator (needs to initialize the store in the ServerSalt and TrafficLimiter)
* fixes two bugs found by Scrutinizer
* update the Jdenticon library (dependabot had only updated the composer.lock, not the actual library code)
* adds a test for the migration script
* discovered that the migration script requires PHP 7.1 for the third getopt parameter, so skipping the test on the unsupported version and added an error message to the script
* small optimizations to `Database`, found while looking at the next item...
* weirdly, the tests on all versions passed before we merged the migration script, but the `ModelTest::testCommentStoreFail` started receiving a different exception under PHP < 7.2 _after_ the change got merged. Clearly, this has to do with the removal of the cache in the `Database` class and I assume that there was some issue with the github actions' cache of phpunit that caused this to only surface after the merge? Tried several things, but in the end gave up and defanged that test when running under PHP < 7.2 - I don't think this is a real world issue, since the behaviour is still correct in that creation of comments can fail even when the paste exists, it just fails in a different area then before (the test drops the comment table to force that issue).

### CODE REVIEW

File: lib/Data/Filesystem.php [hunk 1]
```diff
@@ -409,38 +409,47 @@ function ($filename) {
...
-            }
-            $subsubdirs = preg_grep('/^[^.].$/', $subsubdirs);
-            foreach ($subsubdirs as $subsubdir) {
-                $subsubpath = $subpath . DIRECTORY_SEPARATOR . $subsubdir;
+        $pastes     = array();
+        $firstLevel = array_filter(
+            scandir($this->_path),
+            'PrivateBin\Data\Filesystem::_isFirstLevelDir'
+        );
+        if (count($firstLevel) > 0) {
```
Comment (rugk) [inline lib/Data/Filesystem.php:417]: code style: here begins a huge block of nested and nested and nested code, so it's very hard for me to follow here. (see also: clean code)

I'd suggest to use early return here: https://dev.to/i3uckwheat/early-return-pattern-558i

File: lib/Data/Filesystem.php [hunk 2]
```diff
@@ -409,38 +409,47 @@ function ($filename) {
...
-                        $pastes[] = substr($file, 0, strlen($file) - 4);
+                foreach ($secondLevel as $secondKey) {
+                    $path = $this->_path . DIRECTORY_SEPARATOR . $firstKey .
+                        DIRECTORY_SEPARATOR . $secondKey;
+                    if (!is_dir($path)) {
+                        continue;
                     }
+                    $thirdLevel = array_filter(
+                        array_map(
+                            function ($filename) {
```
Comment (rugk) [inline lib/Data/Filesystem.php:437]: Is that an inline function, uff? Can we at least give that a name and make it a non-anonymous function?

It seems some filename splitting? That could be a helper and also, does it assume a file extension always consists of 4 characters? (If so, that is obviously wrong, we have `.doc` and `.docx` e.g. Also I **hope** PHP has some library or so for doing that…)

So actionable again:
* Split that into own file/class/helper even. It's basic FS functionality. Clean-code wise.
* Maybe we can find a better library that does this task for us? 
* A short search also showed a better way using `explode` (looks better for me without knowing the details here)
  https://stackoverflow.com/a/37856144/5008962
* We can then add a proper parametrized test only for that small unit of splitting file names…

File: tst/ModelTest.php [hunk 3]
```diff
@@ -271,8 +274,9 @@ public function testCommentStoreFail()
         $statement->execute();
         $statement->closeCursor();
 
-        $comment = $paste->getComment(Helper::getPasteId());
-        $comment->setData($commentData);
+        if (version_compare(PHP_VERSION, '7.2.0') < 0) {
```
Comment (rugk) [inline tst/ModelTest.php:277]: This is the last point you talk about in your last PR comment?

I hardly get what is happening here… does not the test also fail if you manually throw the exception now?

Also, if you'd wanted to debug this, did you maybe give a try using `git bisect` to debug/find the commit that caused this change?

(And no, I don't want to imply that this is important, after all if that works in a new version that may be fine, but I cannot judge/decide that.)

File: lib/Data/Filesystem.php (hunk 1 above)
Comment (rugk) [reply lib/Data/Filesystem.php:417]: Also, one may generally extract some functions below into some private functions. E.g. the part in the foreach loop.

Or maybe rewrite this to be recursive (the task of doing that looks like being suitable for that).

Comment (rugk) [reply lib/Data/Filesystem.php:417]: Also, may I ask the heretical question of what was the aim of this refactoring: just code cleanup (for new PHP versions?) or also performance improvements?

In case of the last one: Did it actually work, do we have benchmarks?
And even if it was not the aim, I hope that such refactoring does at least not _decrease_ the performance. Can we assure that?
(Returning all pastes seems to be a performance relevant step and may be a critical bottleneck, that's just why I am asking…)

**Review by rugk — COMMENTED** [review]
BTW what does "bin cleanup" refer to in your PR title? Bin = Trash?  **Edit:** Ah me stupid. The bin folder is meant. :laughing: 

I did my best on this review, but I see my PHP knowledge is really outdated and lacks a bit, so these are more or less only some questions and general style comments…

---

Also, as I read the previous discussion or so I agree it would be great if @felixjogris as the original script author could have a look here, too. (Feel free to just use the GitHub review tools.)

File: lib/Data/Filesystem.php (hunk 1 above)
Comment (elrido) [reply lib/Data/Filesystem.php:417]: Purely wanted to match the style of both functions. We can't do an early return, these actually need to loop through all 3 levels of directories to discover all or up to $batchsize * 10 paste IDs. (Aside: that is one area where a RDBMS clearly beats such a naive files on disk storage)

I do like the idea of a recursive function - if we give it a counter argument that it can decrement on each pass, we could use it for both cases. And in any case we should de-dup that logic. I'll work on that, thanks.

File: lib/Data/Filesystem.php (hunk 2 above)
Comment (elrido) [reply lib/Data/Filesystem.php:437]: It's used inside the filter which only returns valid paste IDs. In order to reuse that validation function, we need to remove the suffixes from the paste files - if they then aren't valid IDs, they simply are filtered out. Yes, this is all designed for the case that an admin ends up having garbage files inside the data directory for some reason.

Ideally we can use an operation that makes use of iterators, so we can stream through the scanned files as they are found. Maybe I can rewrite it using the dir class and it's read. Or we could try and use glob, as we know the exact structure of the directories and files: ab/cd/abcdef0123456789.php, then return the sub-string at the expected offsets.

File: tst/ModelTest.php (hunk 3 above)
Comment (elrido) [reply tst/ModelTest.php:277]: My problem is that in my own PHP 7.1 testing environments I couldn't reproduce this. But it happens, most of times but not always in github actions. I gave up and just disabled the tests on those old versions. I know exactly what code change caused this, but I don't know why this happens in github actions and why it did not get detected in the tests running on that change 604c9318751f71a4922b7afcd98eaf277974170b.

> Clearly, this has to do with the removal of the cache in the Database class and I assume that there was some issue with the github actions' cache of phpunit that caused this to only surface after the merge?

The cache avoided several lookups for the same paste in the database - it ran one query and cached the result across all instances of the models (because the cache in the Database was static), so subsequent calls checking for the existence hit the cache instead - and since the cache didn't outlive each request, we didn't have to worry about cache lifetime or eviction. In this test, we know the paste exists, because we just created it and checked it and still, storing the comment fails because it can't find the paste (by query).

The test is quite artificial (dropping a table half-way through a sequence of operations) and I frankly only added it to cover the code path to exception 70 for the coverage report. It's an edge case, but of course I want our code to still detect and handle that error condition and this test validates that scenario.

File: lib/Data/Filesystem.php (hunk 2 above)
Comment (elrido) [reply lib/Data/Filesystem.php]: New solution doesn't need the filters anymore.

File: lib/Data/Filesystem.php (hunk 1 above)
Comment (elrido) [reply lib/Data/Filesystem.php]: So thanks for poking me on this. Managed to be a great use case for a glob-based iterator. Instead of us having to filter out things that may not be pastes we now only need to ignore the discussion directories. Much less code!

Some explanation: The glob may return the old paste files without .php suffix, the current ones with suffix and the directory with .discussion suffix. We only need to ignore the latter ones, all others are valid paste files and the ones without suffix will get upgraded by the exists() call. I forgot about that logic, but FilesystemTest::testOldFilesGetConverted() started failing when my pattern did yield only items with the .php suffix.

**Review by rugk — COMMENTED** [review]
Yeah looks indeed much better now, thanks! :smile:

File: lib/Data/Filesystem.php [hunk 4]
```diff
@@ -481,28 +418,20 @@ private function _dataid2discussionpath($dataid)
...
-     * @access private
-     * @param  string $element
-     * @return bool
+     * @return \GlobIterator
      */
-    private function _isSecondLevelDir($element)
+    private function _getPasteIterator()
     {
-        return (bool) preg_match('/^[a-f0-9]{2}$/', $element);
+        return new \GlobIterator($this->_path . DIRECTORY_SEPARATOR .
```
[the remark this answers is not in the document]
Comment (elrido) [reply lib/Data/Filesystem.php:428]: applied in 07bc3285aab94814996fcbab61cbb992ee139524

> Note that wildcard patterns are not regular expressions, although they
> are a bit similar.

https://man7.org/linux/man-pages/man7/glob.7.html

File: lib/Data/Filesystem.php [hunk 5]
```diff
@@ -481,28 +418,20 @@ private function _dataid2discussionpath($dataid)
...
-    /**
-     * Check that the given element is a valid second level directory.
-     *
-     * @access private
-     * @param  string $element
-     * @return bool
+     * @return \GlobIterator
      */
-    private function _isSecondLevelDir($element)
+    private function _getPasteIterator()
```
[the remark this answers is not in the document]
Comment (elrido) [reply lib/Data/Filesystem.php:426]: It's pretty much a call to [glob() from the C library](https://man7.org/linux/man-pages/man3/glob.3.html). We may assume that [these 1500ish lines](https://codebrowser.dev/glibc/glibc/posix/glob.c.html) of C code in the GNU libc have gotten optimized over the years. It helps that they don't have to do full regex, but stick to proto-regex patterns, allowing them to descend the directories step by step, similarly to how we did it originally.
